### ISSUE TITLE
Training much slower on Keras 2.0.4 than on Keras 1.2.2

### ISSUE THREAD

**Issue by Blockost** [fchollet/keras#6808]
Hi everyone, 

I recently updated Keras to version 2.0.4 and I saw a big drop in performance compared to Keras 1.2.2.

I'm basically trying to find a mapping between movie titles and movie plot summaries (called synopses) using seq2seq LSTMs. While my model was pretty fast with Keras 1.2.2, it took a lots of time to train in the last version.

    model = Sequential()
    model.add(Embedding(vocab_size, EMBEDDING_DIM, input_length=MAX_LENGTH, mask_zero=True))
    model.add(LSTM(1024, return_sequences=True))
    model.add(LSTM(1024, return_sequences=True))
    model.add(TimeDistributed(Dense(vocab_size, activation='softmax')))
    model.compile(loss='sparse_categorical_crossentropy', optimizer=(Adam()), metrics=['accuracy'])

I did not change anything to this model or to the code in general between Keras 1.2.2 and Keras 2.0.4, trained on a Nvidia Quadro K6000 with Theano up-to-date in both cases and simply ran: `pip uninstall keras` and `pip install keras` to update it.

However, the training is much slower.

Using Keras 1.2.2:
`Epoch 1/10
100/100 [==============================] - 5s - loss: 2.2261 - acc: 0.8375`

Using Keras 2.0.4: 
`Epoch 1/10
100/100 [==============================] - 18s - loss: 3.3189 - acc: 0.8277`

In the example above, the delta is only about a few seconds but when training on the whole dataset, it is much much slower.

Does anyone know what I'm doing wrong and could point me in the right direction ?

**Comment by StefanoD** [comment]
Could you look, if your GPU is really used?

```
watch nvidia-smi
```

**Comment by Blockost** [comment]
Yes, I'm sure my GPU is used to perform the calculations: 
![selection_39](https://cloud.githubusercontent.com/assets/7816661/26659914/6645e910-466c-11e7-9402-2a7cddd657de.jpg)
![selection_40](https://cloud.githubusercontent.com/assets/7816661/26659916/679d094c-466c-11e7-8729-9e75d08e9ca8.jpg)

I was more worried about my model, actually. Maybe there's an optimization I need to apply in Keras 2.0.4 to make it better since I didn't change anything to the code (simply updated Keras' version). Unfortunately, I received no warnings or errors from Keras' APIs.

But the perf drop tells me there's definitely something going on behind the scenes.

**Comment by StefanoD** [comment]
Well, 0% GPU is used.

**Comment by Blockost** [comment]
Maybe starting the training at the same time would have helped... Sorry for that:

![selection_41](https://cloud.githubusercontent.com/assets/7816661/26674050/db480226-46b6-11e7-986c-1dfcd856265f.jpg)

**Comment by StefanoD** [comment]
Did you switched backends? Tensorflow is the new default since version 2.

**Comment by Blockost** [comment]
No I didn,'t. My Keras config is still the same: 

`{
    "image_dim_ordering": "th",
    "epsilon": 1e-07,
    "floatx": "float32",
    "backend": "theano"
}`

Should I switch to Tensorflow ?

**Comment by StefanoD** [comment]
I don't see any reasons to switch to Tensorflow.

**Comment by StefanoD** [comment]
What is you image_data_format? Channel first oder last?

**Comment by Blockost** [comment]
I don't know what this means, but here we are: 
![selection_42](https://cloud.githubusercontent.com/assets/7816661/26676875/7d96c49a-46c1-11e7-851a-6324f97bf9f9.jpg)

**Comment by holli** [comment]
@Blockost how do you run the training? Note that if you are using fit_generator the parameters have changed in Keras 2. You might be running more steps than previously.

```
- model.fit_generator(train_generator, train_gen.n, 1)  (old way)
+ model.fit_generator(train_generator, train_gen.n/train_gen.batch_size, 1) (new way)
```

**Comment by Blockost** [comment]
@holli I don't use `fit_generator() `. I simply use `fit()` and pass my dataset as a whole in parameters

**Comment by curiale** [comment]
Hi @holli , I'm using fit_generator and I couldn't figure out why keras 2 is much slower than 1.2. A pice of my code is 

      # fit the model on the batches generated by datagen.flow()
      if int(keras.__version__.split('.')[0]) < 2:
            # Keras 1.2
            model.fit_generator(datagen.flow(img_train, gt_train,
                                         batch_size=batch_size, shuffle=True,),
                            samples_per_epoch=data_augmentation,
                            nb_epoch=nb_epoch,
                            validation_data=(img_test, gt_test),
                            verbose=1, callbacks=callbacks)
        else:
            # Keras 2.0.4

            # Total number of steps (batches of samples) to yield from generator
            steps_per_epoch = data_augmentation / batch_size

            model.fit_generator(datagen.flow(img_train, gt_train,
                                         batch_size=batch_size, shuffle=True,),
                            steps_per_epoch=steps_per_epoch,
                            epochs=nb_epoch,
                            validation_data=(img_test, gt_test),
                            verbose=1, callbacks=callbacks)


so I suppose that I'm using the fit_generator in the right way. Anyone has experienced something like this?

**Comment by StefanoD** [comment]
_data_augmentation_ is really a weird name. Is this the training sample size?

**Comment by curiale** [comment]
Yes, it is the number of training samples per epoch.

**Comment by StefanoD** [comment]
_data_augmentation_ should be the total number of training samples and not training samples per epoch.

**Comment by sir-avinash** [comment]
I'm experiencing the same issue since my update to the Keras 2 API (even after correctly making changes to migrate from samples_per_epoch to steps_per_epoch). @Blockost Could figure out why this is happening?

**Comment by Blockost** [comment]
Sorry but I don't use Keras anymore. If I remember correctly, the quick fix was to downgrade back to 1.2.2 to save some training time... Sorry for not being able to help much with that.
